pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,244 @@
+        self.restarting = self._current_iteration > 0
+
+
+class CaptureIterableDataset(IterableDataset):
+    """
+    The ``CaptureIterableDataset`` is used to wrap an :class:`torch.utils.data.IterableDataset`.
+    On ``__iter__`` function call   the ``CaptureIterableDataset`` will wrap the wrapped dataset
+        generators into ``FastForwardSampler`` to keep track of progress.
+    On ``__next__`` function call, the ``CaptureIterableDataset`` will return a dictionary containing
+        user data and metadata containing the ``FastForwardSampler`` samplers state_dict.
+    """
+
+    def __init__(self, dataset: IterableDataset, initial_seed: Optional[int] = None):
+        super().__init__()
+        self.dataset = dataset
+        self.state_dict: Optional[Dict[int, Any]] = None
+        self.initial_seed = initial_seed
+        self.samplers: Optional[Dict[str, FastForwardSampler]] = None
+
+    @property
+    def sampler(self) -> Sampler:
+        return self.dataset.sampler
+
+    def load_state_dict(self, state_dict: Dict[int, Any]) -> None:
+        self.state_dict = deepcopy(state_dict)
+
+    def _wrap_generator_samplers(self) -> None:
+        if self.samplers is None:
+            self.samplers = {}
+
+            # access wrapped dataset attributes
+            dataset_dict = self.dataset.__dict__
+
+            # create a tuple of sampler names
+            samplers_names = tuple(v.__class__.__name__ for k, v in dataset_dict.items() if isinstance(v, Sampler))
+
+            # create a dictionary of generator present within the dataset attributes
+            dataset_sampler_generators = {k: v for k, v in dataset_dict.items() if isinstance(v, Generator)}
+
+            # iterate over the generator. If a generator was created from a ``Sampler```,
+            # it will be wrapped into a ``FastForwardSampler``.
+            for (generator_attr_name, generator) in dataset_sampler_generators.items():
+
+                # Generator name have the  the form `SamplerName.__iter__`
+                generator_name = generator.__qualname__.split('.')[0]
+
+                # validate the base generator name matches a sampler name.
+                if any(sampler_name == generator_name for sampler_name in samplers_names):
+
+                    # wrap the generator into a ``FastForwardSampler``
+                    sampler = FastForwardSampler(generator)
+
+                    # if ``CaptureIterableDataset`` was available, the sampler should reload its own state.
+                    if self.state_dict is not None:
+                        sampler.load_state_dict(self.state_dict[generator_attr_name])
+
+                    # store the samplers
+                    self.samplers[generator_attr_name] = sampler
+
+                    # replace generator with the generator from the ``FastForwardSampler``.
+                    dataset_dict[generator_attr_name] = iter(sampler)
```

Comment:
Won't this change the dataset `__dict__` state for the user? Would need to use `__dict__.copy()`

Reply:
Added a deep copy on the Dataset.

Reply:
Yes, it does, but we will add logic to un-wrap it in further PR.